### ISSUE TITLE
Support Windows in Istio Data Plane

### ISSUE THREAD

**Issue by sunjayBhatia** [istio/istio#27893]
**Describe the feature request**
Envoy just published Alpha support for the proxy on Windows. As we move to a Beta/GA release we would like to look forward to supporting Windows in the Istio Data Plane as delivering service mesh capabilities for Windows-based applications is something we have significant user signal for. This issue will hopefully be a starting place for Windows support in Istio/Envoy sidecars/data plane.

Where do the maintainers think is the best place to get started with contributions?

While there may be integrations with dependencies (windows, k8s, etc.) that need to get worked out, my initial broad-strokes guess is we need to get the Istio fork of Envoy and `istio/proxy` building/running tests in CI and then start enabling integration testing for Windows service proxies.

[ ] Docs
[ ] Installation
[x] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[x] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Additional context**
We have tried out the Istio Envoy fork and ensured the upcoming 1.8 release branch builds on Windows, however that was without applying the custom extensions from the `istio/proxy` repo.

Links:
- Envoy Windows dev image: https://hub.docker.com/r/envoyproxy/envoy-windows-dev/tags
- Envoy blog posts:
  - https://blog.envoyproxy.io/announcing-alpha-support-for-envoy-on-windows-d2c53c51de7b
  - https://blog.envoyproxy.io/envoy-proxy-on-windows-containers-193dffa13050

**Comment by howardjohn** [comment]
My comment on https://github.com/istio/istio/issues/21094#issuecomment-598787617 applies here as well

Aside from the build, we also need to figure out the sidecar traffic capture since we will not have iptables

**Comment by howardjohn** [comment]
Maybe a dumb question, but can we have mixed windows and linux pod? Like a user deploys their windows workload, but we inject the linux container?

**Comment by sunjayBhatia** [comment]
> My comment on [#21094 (comment)](https://github.com/istio/istio/issues/21094#issuecomment-598787617) applies here as well
> 
> Aside from the build, we also need to figure out the sidecar traffic capture since we will not have iptables

Very good point, not sure if it is publicly noted anywhere yet outside of envoy discussion but our friends at Microsoft are planning on releasing a windows feature in one of the 2021 semi annual releases with an equivalent feature

**Comment by sunjayBhatia** [comment]
> Maybe a dumb question, but can we have mixed windows and linux pod? Like a user deploys their windows workload, but we inject the linux container?

As far as I remember I don’t think this is possible. I dont *think* windows containers and Linux containers on windows can do shared networking. Linux containers on windows also use the more heavyweight hyper-v isolation I believe as well

**Comment by sunjayBhatia** [comment]
cc @daschott

**Comment by sunjayBhatia** [comment]
From the conversation with the Environments WG today, started poking around with `pilot-agent`

Interestingly, the `pilot-agent` program all compiles except for [this code](https://github.com/istio/istio/blob/fd796274081ee9ad26bfd3e77eff4516511d7afb/tools/istio-iptables/pkg/validation/linux.go) (which of course is Linux specific), though it is clear functionally it is not fully operational on Windows. We would need to unwind some of the signal handling/shutdown semantics, filepaths, envoy hot restart usage, iptables usage, etc.

Another little spike/experiment that will be interesting is to see how much of `istio/proxy` works on Windows.

**Comment by howardjohn** [comment]
https://github.com/istio/istio/blob/fd796274081ee9ad26bfd3e77eff4516511d7afb/tools/istio-iptables/pkg/validation/linux.go is really fairly optional. Its only used in the CNI code, where we need to assert the iptables results are set up properly due to some edge cases where it fails. However everything else in istio-iptables/ aside from validation/ is basically shelling out to `iptables` which will not work without some efforts I would imagine.

I would think you could probably get a gateway running (Envoy+pilot-agent, but none of the iptables code is run) without too much issue. You can use an upstream envoy `envoy` binary initially for testing if building istio-proxy is challenging as well (although some features will not work - good enough for a demo/POC)

**Comment by sdake** [comment]
@sunjayBhatia I have placed this in scope for 1.9, although it may be pushed back depending on your level of development. The community would require an RFC to proceed with a merge of such work, although the RFC process itself is not too difficult. You may find it a little difficult though, as a new contributor. If you could do me a solid and get one PR merged - this would allow us to add you as a member of| the Istio organization such that other maintainers can assign the issues related to this work to you.

Cheers,
-steve

**Comment by sdake** [comment]
@sunjayBhatia for pilot-agent, there are a couple of approaches that may work. One would be to build an iptables "shim" - something that simulates iptables but on windows. Another would be to implement a windows version of pilot-agent. I don't know how feasible this second proposal is. Most of the difficulty you will face, is as John has pointed out, focused around iptables.

The remainder is typical porting work of paths and the like.

Finally I wanted to point out - we use a build container and istio/envoy is enabled. It probably doesn't work for Windows, but may get you moving.

```
git clone https://github.com/istio/proxy
BUILD_WITH_CONTAINER=1
make build
make eportcache
```

Again, upstream envoy may be better initially but I just want to point out this functionality.

**Comment by sunjayBhatia** [comment]
@sdake Thanks for the pointers, I will try to get an RFC document out soon and get a small PR going. Is there a specific working group to present the RFC to? (I missed the Networking WG meeting this week due to a conflict but was planning on attending next week)

**Comment by sdake** [comment]
I think an RFC to the environments wg using the community template in CONTRIBUTING.md would be a good start. The template has embedded instructions.

Based upon scope, the WG can determine who else to include in the RFC signoff.

Cheers,
-steve

**Comment by sunjayBhatia** [comment]
Cross-linking RFC and enhancement issue:
- https://docs.google.com/document/d/1nlEeo_PDoZtb9MHB1L9JpnbdU8efpcGJwQDwlWz7LXk/edit?usp=sharing
- https://github.com/istio/enhancements/issues/10

**Comment by Type1J** [comment]
> From the conversation with the Environments WG today, started poking around with `pilot-agent`
> 
> Interestingly, the `pilot-agent` program all compiles except for [this code](https://github.com/istio/istio/blob/fd796274081ee9ad26bfd3e77eff4516511d7afb/tools/istio-iptables/pkg/validation/linux.go) (which of course is Linux specific), though it is clear functionally it is not fully operational on Windows. We would need to unwind some of the signal handling/graceful shutdown semantics, filepaths, envoy hot restart usage, iptables usage, etc.
> 
> Another little spike/experiment that will be interesting is to see how much of `istio/proxy` works on Windows.

Please correct me if I'm wrong, but being that most (all?) Kubernetes clusters have at least 1 Linux node, does Pilot even need a port? It might be interesting to get all of Istiod working on Windows, but I don't think it's nessisary. I think that focusing on the sidecar injected into Windows pods will get us to a good place: Windows pods and Linux pods working together with Istio.

**Comment by sunjayBhatia** [comment]
> does Pilot even need a port

The goal here was not to port all of Pilot but just the `pilot-agent` which runs in a sidecar/init container and bootstraps Envoy and sets up traffic capture for the application pod (https://istio.io/latest/docs/reference/commands/pilot-agent/)

> Kubernetes clusters have at least 1 Linux node

This is true, see https://kubernetes.io/docs/setup/production-environment/windows/intro-windows-in-kubernetes/#windows-containers-in-kubernetes

**Comment by Type1J** [comment]
@sunjayBhatia Thanks for the clairification.

**Comment by Type1J** [comment]
I have many projects that would benefit from Istio, but their main workload containers are on Windows. As soon as this is available, I plan to use it on all new projects, and maybe retrofit a few as well. Keep up the good work @sunjayBhatia!

**Comment by sunjayBhatia** [comment]
Cross-referencing comment: https://github.com/istio/enhancements/issues/10#issuecomment-758038914

Currently this work has been deprioritized, however happy to pass along ownership to any interested parties

**Comment by joeapearson** [comment]
@sunjayBhatia very possibly we may be able to make some contributions in this regard.  Do you have anything shareable at this point?

**Comment by sunjayBhatia** [comment]
> @sunjayBhatia very possibly we may be able to make some contributions in this regard. Do you have anything shareable at this point?

Not much in the way of code, there was mainly only spike work done, this document was the main output of what was worked on: https://docs.google.com/document/d/1nlEeo_PDoZtb9MHB1L9JpnbdU8efpcGJwQDwlWz7LXk/edit?usp=sharing

**Comment by howardjohn** [comment]
Not stale

**Comment by desek** [comment]
Envoy is now GA for Windows: https://blog.envoyproxy.io/general-availability-of-envoy-on-windows-267e4544994a

**Comment by Hkyx** [comment]
@desek the thread have 6months, do we have a official release now instead of a alpha?

**Comment by Type1J** [comment]
I'd like to know, as well: Does Istio now support Windows merely by Envoy running in Windows?

**Comment by howardjohn** [comment]
No, running envoy is only part of the problem to solve. Still need to make
sure all the Istio binaries are built and work on windows. Namely, the
redirection stuff will need to be bespoke for windows.

**Comment by Type1J** [comment]
Does ztunnel get us any closer to Windows support?

**Comment by howardjohn** [comment]
Sort of? It is a smaller surface to port to windows if we did chose to. But it doesn't support windows currently.

**Comment by Type1J** [comment]
Please make ztunnel amd wayport work ok Windows!

**Comment by howardjohn** [comment]
Just being realistic, its not going to happen unless someone that is a windows expert decides to invest a lot of time into it. Its not a priority for the project nor something anyone currently working on Istio has expertise in.

Also waypoints are probably a non-goal; they can just  run on Linux while the ztunnel runs on windows.

**Comment by vrapolinario** [comment]
What type of support would you need on this? While I don't think the Windows containers team would be able to jump in here to contribute (development wise) there's definitely the possibility of technical support. I'd be happy to make connections and help myself where I can.

**Comment by howardjohn** [comment]
Basically all the redirection logic to transparently redirect traffic to the ztunnel. and then all the TPROXY stuff in ztunnel itself

**Comment by Type1J** [comment]
If these components can run on a Linux machine, and Windows containers' traffic could be transparently proxied through them, then that would solve the issue, I believe. Sure, some security is sacrificed (mTLS may not be used between the Linux and Windows node), but traffic would be monitorable and comprimised Pods would be detectable.

**Comment by keithmattix** [comment]
Even in ambient, the ztunnels would have to be able to run on Windows nodes unfortunately.

**Comment by Type1J** [comment]
> Even in ambient, the ztunnels would have to be able to run on Windows nodes unfortunately.

That's what I thought from what I've seen. The questions are:

1. Does ztunnel do anything that can't be done on Windows?
2. Can waypoint on another node be used for the layer 7 functionality?

**Comment by howardjohn** [comment]
> 1. Does ztunnel do anything that can't be done on Windows?

Yes, there are a few things.

1. Redirection - this is not actually done by ztunnel but by the CNI daemonset, but still needs to happen
2. original dst and tproxy. Also `splice` but that is optional
3. Some signal stuff

Most/all can be found with:

```
$ rg 'cfg.*unix|cfg.*linux'
src/signal.rs
66:#[cfg(unix)]
96:#[cfg(not(unix))]

src/socket.rs
22:#[cfg(target_os = "linux")]
30:#[cfg(target_os = "linux")]
35:#[cfg(target_os = "linux")]
74:#[cfg(target_os = "linux")]
97:#[cfg(not(target_os = "linux"))]
105:#[cfg(not(target_os = "linux"))]
113:#[cfg(not(target_os = "linux"))]
121:#[cfg(target_os = "linux")]
155:#[cfg(all(target_os = "linux"))]
171:#[cfg(not(target_os = "linux"))]
```

Just to clarify , I mean _does not today_. I am sure its theoretically possible to replace these on windows.

> 2. Can waypoint on another node be used for the layer 7 functionality?

Yes, that is one benefit of the architecture - waypoint can run anywhere

**Comment by AndreaPQ** [comment]
Traffic redirection on Windows 
https://github.com/microsoft/hcnproxyctrl
On Linux, iptables rules are used to intercept and redirect traffic. On Windows, this is achieved using the HNS [L4WfpProxyPolicySetting](https://docs.microsoft.com/en-us/virtualization/api/hcn/hns_schema#l4wfpproxypolicysetting). The L4WfpProxyPolicySetting utilizes the [Windows Filtering Platform](https://docs.microsoft.com/en-us/windows/win32/fwp/about-windows-filtering-platform) (WFP) for network traffic filtering and redirection

See Linkerd2 Issue "Add Support For Hybrid Windows/Linux Clusters"
https://github.com/linkerd/linkerd2/issues/3298#issuecomment-1617694226

**Comment by keithmattix** [comment]
Hey folks! If you've been following this 4.5 year old thread, then first of all, thanks for your patience! We've made some substantial progress in this area, and we're excited to hear more about the needs and use-cases in this space. You can follow along with that progress in the two PRs adding the initial bit of functionality: istio/ztunnel#1461 and #55216. Feel free to reach out to me directly as well

**Comment by it2008018** [comment]
Hi @keithmattix ,
Thanks you for your extensive work on this. By when we should expect this to get merged and available for test/use?

**Comment by keithmattix** [comment]
We hope to get it merged into an experimental branch by 1.26 or 1.27

**Comment by heyjustarooni** [comment]
Hey @keithmattix thank you for your work on this! I'm in the process of building out a Windows/Linux RKE2 cluster and would love to be able to use Istio Ambient for everything. As of now we are working on standing up Traefik alongside Istio as a stop gap which unfortunately adds more complexity to our setup. Can't wait to start testing the new Windows support!

**Comment by ctradar** [comment]
@keithmattix can we use this windows experimental branch to test it out internally ?

https://github.com/istio/istio/tree/experimental-windows-ambient

I see 1.26.0 has been released last week. Is Windows support pushed to be included in 1.27 release ?

**Comment by keithmattix** [comment]
Yes it was pushed to 1.27. Need some more time to get the build process aligned with Istio's

**Comment by grnmeira** [comment]
The Windows work (ambient only for now), is being tracked by the windows/ambient tag. Same for required ZTunnel work.

**Comment by pavelpikatamcs** [comment]
> Yes it was pushed to 1.27. Need some more time to get the build process aligned with Istio's

Is there any update for this objective? 1.28 is out now and still no Windows support :(

**Comment by ShadowLNC** [comment]
@keithmattix @grnmeira can you please confirm if is this is still on the near-term roadmap?

**Comment by DWilson-Soloio** [comment]
@keithmattix @grnmeira can you provide any more detail regarding release this will be in?

**Comment by wmorgan** [comment]
The enterprise version of Linkerd now has first-class Windows support (both VMs/bare metal and Windows containers) if anyone is open to commercial alternatives. https://docs.buoyant.io/buoyant-enterprise-linkerd/latest/features/windows-mesh-expansion/

**Comment by grnmeira** [comment]
@DWilson-Soloio, @ShadowLNC  Windows support is still experimental, there's no releases planned for it so far. @zirain I believe has done the most recent work on it. Let us know if you have any info.
